stdlib/public/core/StringStorage.swift

```diff
@@ -93,102 +281,88 @@ final internal class __StringStorage
 #else
-    storage._realCapacityAndFlags =
-      UInt64(truncatingIfNeeded: realCodeUnitCapacity)
+    storage._capacityAndFlags = capAndFlags
     storage._countAndFlags = countAndFlags
 #endif
 
-    storage._breadcrumbsAddress.initialize(to: nil)
+    _internalInvariant(
+      storage._countAndFlags._storage == countAndFlags._storage)
+    _internalInvariant(
+      storage._capacityAndFlags._storage == capAndFlags._storage)
+    _internalInvariant(
+      storage.unusedCapacity == capAndFlags.capacity - countAndFlags.count)
+
+    if storage.hasBreadcrumbs {
+      storage._breadcrumbsAddress.initialize(to: nil)
+    }
+
     storage.terminator.pointee = 0 // nul-terminated
 
-    // NOTE: We can't _invariantCheck() now, because code units have not been
-    // initialized. But, _StringGuts's initializer will.
+    // We can check layout invariants, but our code units have not yet been
+    // initialized so we can't verify e.g. ASCII-ness
+    storage._invariantCheck(initialized: false)
     return storage
   }
 
-  @_effects(releasenone)
-  private static func create(
-    capacity: Int, countAndFlags: CountAndFlags
+  // DO NOT PUSH: For stats gathering only...
+  internal static func create(
+    count: Int, codeUnitCapacity capacity: Int
   ) -> __StringStorage {
-    _internalInvariant(capacity >= countAndFlags.count)
-
-    let realCapacity = determineCodeUnitCapacity(capacity)
-    _internalInvariant(realCapacity > capacity)
-    return __StringStorage.create(
-      realCodeUnitCapacity: realCapacity, countAndFlags: countAndFlags)
+    __StringStorage.create(
+      codeUnitCapacity: capacity,
+      countAndFlags: _CountAndFlags(mortalCount: count, isASCII: false))
+    // NOTE: content is garbage and not necessarily validy endcoded. Please
+    // delete this entry point...
   }
 
   // The caller is expected to check UTF8 validity and ASCII-ness and update
   // the resulting StringStorage accordingly
   internal static func create(
-    uninitializedCapacity capacity: Int,
+    uninitializedCodeUnitCapacity capacity: Int,
     initializingUncheckedUTF8With initializer: (
       _ buffer: UnsafeMutableBufferPointer<UInt8>
     ) throws -> Int
   ) rethrows -> __StringStorage {
     let storage = __StringStorage.create(
-      capacity: capacity,
-      countAndFlags: CountAndFlags(mortalCount: 0, isASCII: false)
+      codeUnitCapacity: capacity,
+      countAndFlags: _CountAndFlags(mortalCount: 0, isASCII: false)
     )
     let buffer = UnsafeMutableBufferPointer(start: storage.mutableStart,
                                             count: capacity)
     let count = try initializer(buffer)
 
-    let countAndFlags = CountAndFlags(mortalCount: count, isASCII: false)
+    let countAndFlags = _CountAndFlags(mortalCount: count, isASCII: false)
     #if arch(i386) || arch(arm) || arch(wasm32)
     storage._count = countAndFlags.count
-    storage._flags = countAndFlags.flags
+    storage._countFlags = countAndFlags.flags
```

Comment:
having count, countFlags, and countAndFlags on the same thing is a bit confusing heh